### PR TITLE
Better waiting for watch event delivery in cacher

### PR DESCRIPTION
**Pull request by wojtek-t**
@lavalamp - I think we should do something simple for now (and merge for 1.5), and do something a bit more sophisticated right after 1.5, WDYT?

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37195)

### CODE REVIEW

File: pkg/storage/cacher.go [hunk 1]
```diff
@@ -531,17 +531,23 @@ func (c *Cacher) dispatchEvents() {
 func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	triggerValues, supported := c.triggerValues(event)
 
+	// TODO: For now we assume we have a given <timeout> budget for dispatching
+	// a single event. We should consider changing to the approach with:
+	// - budget has upper bound at <max_timeout>
+	// - we add <portion> to current timeout every second
+	timeout := time.Duration(250) * time.Millisecond
```
Comment (lavalamp) [inline pkg/storage/cacher.go]: does dispatchEvent get called more than 4 times per second?

File: pkg/storage/cacher.go [hunk 2]
```diff
@@ -768,6 +770,10 @@ func (c *cacheWatcher) add(event *watchCacheEvent) {
 		c.forget(false)
 		c.stop()
 	}
+
+	if *timeout = *timeout - time.Since(startTime); *timeout < 0 {
```
Comment (lavalamp) [inline pkg/storage/cacher.go]: I would probably extract this and the timerPool bit to a separate function / object, just for clarity & to make testing easier. Up to you.

File: pkg/storage/cacher.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: Yup - in large clusters it can be even hundreds (e.g. in 2k-node clusters, it can be 250 or so).
But as I wrote somewhere, it happens that some watcher sometimes block it for tens of milliseconds (e.g. 20 or 30) and system still works perfectly fine. Cancelling them would potentially cause relists, which might be even worse for the system.
Maybe 250ms is too large, but I think we shouldn't get lower than 100ms (at least now).

File: pkg/storage/cacher.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: Can we do it after 1.5 is out (while doing something more sophisticated)?

Comment (lavalamp) [inline pkg/storage/cacher.go]: sure.

File: pkg/storage/cacher.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: It's not a full fix, but it would solve the main underlying problem (there won't be hanging for few minutes List requests - because we will be releasing lock at least once every ~250ms). I verified that indeed that one is fixed and extended some test. It doesn't protect for really stale data in watch, but I think it's slightly too late for such fix now. I will try to do it next week.

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE (build started)

**Comment by wojtek-t** [comment]
Verify failed twice with
"I1127 05:06:59.517] FAILED   hack/make-rules/../../hack/verify-federation-openapi-spec.sh	102s"

I rebased and pushed the new version (I'm not able to repro it locally).

**Comment by wojtek-t** [comment]
@k8s-bot kubemark test this, issue: #IGNORE (Waiting for status to be reported)

**Comment by wojtek-t** [comment]
@k8s-bot kubemark test this, issue: #IGNORE (Error in queue.)

**Comment by dims** [comment]
@k8s-bot test this

**Comment by dims** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this, issue: #IGNORE (error starting build)

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE (error in queueing build)

**Comment by wojtek-t** [comment]
@k8s-bot verify test this, issue: #IGNORE (Error queueing build.)

**Comment by wojtek-t** [comment]
@k8s-bot kops test this, issue: #IGNORE (Error queueing build.)

**Comment by smarterclayton** [comment]
@liggitt changes the dynamics of how watches are delivered under load - may have impact for our UI in some cases (because we have so many more watches)

**Comment by liggitt** [comment]
can you summarize how this impacts watch event delivery in the description? as an API consumer, are watch events less reliable now?

**Comment by wojtek-t** [comment]
@liggitt - they are more reliable - before that with large numbers of watchers not receiving events, it may have taken minutes to deliver a single watch event (and we observed it). With it there are some guarantees (up to ~250ms). A more sophisticate fix will come soon (though probably won't be cherrypicked to 1.5.0).

**Comment by liggitt** [comment]
what gives if the 250ms timeout is hit when sending? does that mean a slow watcher drops events if it gets more than 250ms behind?

**Comment by wojtek-t** [comment]
@liggitt - each watcher has a buffer; if we cannot add anything to the buffer, then the watcher is stopped and it needs to be renewed by the client. the watcher should never drop events (both before and after this change)
